runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -1347,22 +1943,50 @@ TR::Register *J9::Power::TreeEvaluator::irdbariEvaluator(TR::Node *node, TR::Cod
    cg->stopUsingRegister(condReg);
 
    cg->machine()->setLinkRegisterKilled(true);
-
-   return objReg;
-#else
-   return NULL;
+   return objReg;   
 #endif
    }
 
 TR::Register *J9::Power::TreeEvaluator::ardbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return TR::TreeEvaluator::ardbariEvaluator(node, cg);
+   // For rdbar and wrtbar nodes we first evaluate the children we need to
+   // handle the side effects. Then we delegate the evaluation of the remaining
+   // children and the load/store operation to the appropriate load/store evaluator.
+   TR::Node *sideEffectNode = node->getFirstChild();
+   TR::Register *sideEffectRegister = cg->evaluate(sideEffectNode);
+   if (cg->comp()->getOption(TR_EnableFieldWatch))
+      {
+      TR::TreeEvaluator::rdWrtbarHelperForFieldWatch(node, cg, sideEffectRegister, NULL);
+      }
+   cg->decReferenceCount(sideEffectNode);
+   return TR::TreeEvaluator::aloadEvaluator(node, cg);
    }
 
 TR::Register *J9::Power::TreeEvaluator::ardbariEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR_ASSERT(TR::Compiler->om.readBarrierType() != gc_modron_readbar_none, "aReadBarrierEvaluator");
-#ifdef OMR_GC_CONCURRENT_SCAVENGER
+   // For rdbar and wrtbar nodes we first evaluate the children we need to
+   // handle the side effects. Then we delegate the evaluation of the remaining
+   // children and the load/store operation to the appropriate load/store evaluator.
+   TR::Register *sideEffectRegister = cg->evaluate(node->getFirstChild());
+   if (cg->comp()->getOption(TR_EnableFieldWatch))
+      {
+      TR::TreeEvaluator::rdWrtbarHelperForFieldWatch(node, cg, sideEffectRegister, NULL);
+      }
+   // Note: For indirect rdbar nodes, the first child (sideEffectNode) is also used by the 
+   // load evaluator. The load evaluator will also evaluate+decrement it. In order to avoid double
+   // decrementing the node we skip doing it here and let the load evaluator do it.
+   if (TR::Compiler->om.readBarrierType() == gc_modron_readbar_none)
+      return TR::TreeEvaluator::aloadEvaluator(node, cg);
+   else
+      return TR::TreeEvaluator::aConcurrentScavengeSequence(node, cg);
+   }
+
+TR::Register *J9::Power::TreeEvaluator::aConcurrentScavengeSequence(TR::Node *node, TR::CodeGenerator *cg)
```

Comment:
Can we inline this as well? Similar to the `i` variant, We only invoke this routine once and it makes sense that it's a part of the evaluator itself.

Reply:
Actually, I am not sure I agree. Putting it into separate function makes the logic more clear in the caller, eg. we are either calling aloadEvaluator or aConcurrentScavengeSequence -- more symmetrical.

Reply:
That's a good point. You're right, in this case it makes sense that we place the logic in a different routine so that it's more symmetrical and thus easy to understand.